pandas/core/arrays/arrow/accessors.py

```diff
@@ -314,27 +336,90 @@ def field(self, name_or_index: str | int) -> Series:
         1    2
         2    1
         Name: version, dtype: int64[pyarrow]
+
+        Or an expression
+
+        >>> import pyarrow.compute as pc
+        >>> s.struct.field(pc.field("project"))
+        0    pandas
+        1    pandas
+        2     numpy
+        Name: project, dtype: string[pyarrow]
+
+        For nested struct types, you can pass a list of values to index
+        multiple levels:
+
+        >>> version_type = pa.struct([
+        ...     ("major", pa.int64()),
+        ...     ("minor", pa.int64()),
+        ... ])
+        >>> s = pd.Series(
+        ...     [
+        ...         {"version": {"major": 1, "minor": 5}, "project": "pandas"},
+        ...         {"version": {"major": 2, "minor": 1}, "project": "pandas"},
+        ...         {"version": {"major": 1, "minor": 26}, "project": "numpy"},
+        ...     ],
+        ...     dtype=pd.ArrowDtype(pa.struct(
+        ...         [("version", version_type), ("project", pa.string())]
+        ...     ))
+        ... )
+        >>> s.struct.field(["version", "minor"])
+        0     5
+        1     1
+        2    26
+        Name: minor, dtype: int64[pyarrow]
+        >>> s.struct.field([0, 0])
+        0    1
+        1    2
+        2    1
+        Name: major, dtype: int64[pyarrow]
         """
         from pandas import Series
 
+        def get_name(
+            level_name_or_index: list[str]
+            | list[bytes]
+            | list[int]
+            | pc.Expression
+            | bytes
+            | str
+            | int,
+            data: pa.ChunkedArray,
+        ):
+            if isinstance(level_name_or_index, int):
+                index = data.type.field(level_name_or_index).name
+            elif isinstance(level_name_or_index, (str, bytes)):
+                index = level_name_or_index
+            elif isinstance(level_name_or_index, pc.Expression):
+                index = str(level_name_or_index)
+            elif is_list_like(level_name_or_index):
```

Comment:
This is a bit more complicated than I would have hoped :/

It also feels really fragile when combined `pc.Expression` objects that aren't simply `pc.field(name)`. It *might* not be worth supporting name extraction for list-like inputs.

Reply:
`pc.field(0)` is an example of some weirdness:

```python
In [26]: s
Out[26]:
0    {'version': {'major': 1, 'minor': 5}, 'project...
1    {'version': {'major': 2, 'minor': 1}, 'project...
2    {'version': {'major': 1, 'minor': 26}, 'projec...
dtype: struct<version: struct<major: int64, minor: int64>, project: string>[pyarrow]

In [27]: s.struct.field(pc.field(0))
Out[27]:
0     {'major': 1, 'minor': 5}
1     {'major': 2, 'minor': 1}
2    {'major': 1, 'minor': 26}
Name: FieldPath(0), dtype: struct<major: int64, minor: int64>[pyarrow]
```

At the moment, it seems like `pc.struct_field` doesn't support a list of expressions:

```pytb
In [33]: pc.struct_field(s.array._pa_array, [pc.field(0)])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[33], line 1
----> 1 pc.struct_field(s.array._pa_array, [pc.field(0)])

File ~/gh/pandas-dev/pandas/.direnv/python-3.11/lib/python3.11/site-packages/pyarrow/compute.py:258, in _make_generic_wrapper.<locals>.wrapper(memory_pool, options, *args, **kwargs)
    256 else:
    257     option_args = ()
--> 258 options = _handle_options(func_name, options_class, options,
    259                           option_args, kwargs)
    260 if args and isinstance(args[0], Expression):
    261     return Expression._call(func_name, list(args), options)

File ~/gh/pandas-dev/pandas/.direnv/python-3.11/lib/python3.11/site-packages/pyarrow/compute.py:221, in _handle_options(name, options_class, options, args, kwargs)
    216     if options is not None:
    217         raise TypeError(
    218             "Function {!r} called with both an 'options' argument "
    219             "and additional arguments"
    220             .format(name))
--> 221     return options_class(*args, **kwargs)
    223 if options is not None:
    224     if isinstance(options, dict):

File ~/gh/pandas-dev/pandas/.direnv/python-3.11/lib/python3.11/site-packages/pyarrow/_compute.pyx:1474, in pyarrow._compute.StructFieldOptions.__init__()

File ~/gh/pandas-dev/pandas/.direnv/python-3.11/lib/python3.11/site-packages/pyarrow/_compute.pyx:1458, in pyarrow._compute._StructFieldOptions._set_options()

File ~/gh/pandas-dev/pandas/.direnv/python-3.11/lib/python3.11/site-packages/pyarrow/_compute.pyx:1428, in pyarrow._compute._ensure_field_ref()

File ~/gh/pandas-dev/pandas/.direnv/python-3.11/lib/python3.11/site-packages/pyarrow/_compute.pyx:2567, in pyarrow._compute.Expression._nested_field()

File stringsource:15, in string.from_py.__pyx_convert_string_from_py_std__in_string()
```

But if it did, then this `get_name` function would fail because the repr of the expression doesn't equal the name of the field from the type.

Reply:
I'd be okay with omitting name extraction. Maybe for list inputs we say we do `".".join(str(part) for part in name_or_index)`. That's a bit inconsistent since a list of integer indexes would end up with names like `0.1.0`, but at least it's predictable.

Reply:
I think that might be best.

Reply:
Actually, I take this back. We *do* handle the `pc.field(0)` case correctly:

```
ipdb> pp name_or_index
[<pyarrow.compute.Expression FieldPath(0)>]
ipdb> pp get_name(name_or_index, pa_arr)
'string_col'
```

So I think we're good as is.